com.unity.render-pipelines.core/Runtime/Debugging/ShaderDebugPrintManager.cs

```diff
@@ -0,0 +1,323 @@
+using System;
+using System.Collections.Generic;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using UnityEngine.Assertions;
+
+namespace UnityEngine.Rendering
+{
+    public sealed class ShaderDebugPrintManager
+    {
+        private static readonly ShaderDebugPrintManager instance = new ShaderDebugPrintManager();
+
+        private const int DebugUAVSlot = 7;
+        private const int FramesInFlight = 4;
+        private const int MaxBufferElements = 1024 * 16; // Must match the shader size definition
+
+        private List<GraphicsBuffer> m_OutputBuffers = new List<GraphicsBuffer>();
+
+        private List<Rendering.AsyncGPUReadbackRequest> m_readbackRequests =
+            new List<Rendering.AsyncGPUReadbackRequest>();
+
+        // Cache Action to avoid delegate allocation
+        private Action<AsyncGPUReadbackRequest> m_bufferReadCompleteAction;
+
+        private int m_FrameCounter = 0;
+        private bool m_FrameCleared = false;
+
+        private string m_OutputLine = "";
+        private Action<string> m_OutputAction;
+
+        private static readonly int m_ShaderPropertyIDInputMouse = Shader.PropertyToID("_ShaderDebugPrintInputMouse");
+        private static readonly int m_ShaderPropertyIDInputFrame = Shader.PropertyToID("_ShaderDebugPrintInputFrame");
+
+        enum DebugValueType
+        {
+            TypeUint = 1,
+            TypeInt = 2,
+            TypeFloat = 3,
+            TypeUint2 = 4,
+            TypeInt2 = 5,
+            TypeFloat2 = 6,
+            TypeUint3 = 7,
+            TypeInt3 = 8,
+            TypeFloat3 = 9,
+            TypeUint4 = 10,
+            TypeInt4 = 11,
+            TypeFloat4 = 12,
+            TypeBool = 13,
+        };
+
+        private const uint TypeHasTag = 128;
+
+        static ShaderDebugPrintManager()
+        {
+        }
+
+        private ShaderDebugPrintManager()
+        {
+            for (int i = 0; i < FramesInFlight; i++)
+            {
+                m_OutputBuffers.Add(new GraphicsBuffer(GraphicsBuffer.Target.Structured, MaxBufferElements, 4));
+                m_readbackRequests.Add(new Rendering.AsyncGPUReadbackRequest());
+            }
+
+            m_bufferReadCompleteAction = BufferReadComplete;
+            m_OutputAction = DefaultOutput;
+        }
+
+        public static ShaderDebugPrintManager Instance
+        {
+            get { return instance; }
+        }
```

Comment:
See my comment with using `Lazy`, besides that properties must start with lowercase